File: basic/src/main/scala/io/github/mimoguz/layeredfonticon/LayeredFontIcon.scala
package io.github.mimoguz.layeredfonticon

import java.awt.{Color, Component, Font}
import javax.swing.{GrayFilter, Icon, JComponent, UIManager}

class LayeredFontIcon(_font: Font, _layers: Seq[Layer]) extends LayeredFontIconBase(_font, _layers):
  override protected def disabledColorFunction: Color => Color =
    val grayFilter = GrayFilter(false, 50) // TODO: Default gray filter?
    c => Color(grayFilter.filterRGB(0, 0, c.getRGB))

  override protected def selectedForeground(component: Component, default: => Color): Color = default

object LayeredFontIcon:
  def apply(font: Font, layers: Layer*): LayeredFontIcon = new LayeredFontIcon(font, layers)

  def apply(
      font: Font,
      symbol: String,
      color: LayerColor = LayerColor.Unset,
      forceColor: Boolean = false,
      xOffset: Float = 0f,
      yOffset: Float = 0f,
  ): LayeredFontIcon = new LayeredFontIcon(font, Seq(Layer(symbol, color, forceColor, xOffset, yOffset)))

  def of(font: Font, symbol: String): LayeredFontIcon = apply(font, Layer(symbol))
  def of(font: Font, layers: Array[Layer]): LayeredFontIcon = apply(font, layers*)
end LayeredFontIcon

File: src/layeredfonticon/src/main/java/io/github/mimoguz/layeredfonticon/LayeredFontIcon.java
package io.github.mimoguz.layeredfonticon;

import javax.swing.GrayFilter;
import java.awt.Color;
import java.awt.Component;
import java.awt.Font;
import java.util.function.Supplier;

/**
 * Basic implementation of {@link LayeredFontIconBase}.
 *
 * @see LayeredFontIconBase
 */
public class LayeredFontIcon extends LayeredFontIconBase {

    public LayeredFontIcon(Font font, Layer[] layers) {
        super(font, layers);
    }

    @Override
    protected Color getDisabledColor(Color foreground) {
        return new Color(grayFilter.filterRGB(0, 0, foreground.getRGB()));
    }

    @Override
    protected Color getSelectedForegroundColor(Component component, Supplier<Color> defaultColor) {
        return defaultColor.get();
    }

    private static GrayFilter grayFilter = new GrayFilter(true, 50);

    /**
     * Sets the shared gray filter used to gray out the icon when it is disabled.
     *
     * @param filter The gray filter.
     */
    public static void setGrayFilter(GrayFilter filter) {
        grayFilter = filter;
    }

    /**
     * A convenience method to create a {@link LayeredFontIcon} instance from layers given as varargs.
     *
     * @param font   The font to use.
     * @param layers The layers of the icon.
     * @return A new {@link LayeredFontIcon} instance.
     */
    public static LayeredFontIcon of(Font font, Layer... layers) {
        return new LayeredFontIcon(font, layers);
    }

    /**
     * A convenience method to create a basic, single-layer {@link LayeredFontIcon} instance.
     *
     * @param font   The font to use.
     * @param symbol The symbol string.
     * @return A new {@link LayeredFontIcon} instance.
     */
    public static LayeredFontIcon of(Font font, String symbol) {
        return LayeredFontIcon.of(font, Layer.of(symbol));
    }

}
